app/src/main/java/org/oppia/app/parser/StringToFractionParser.kt

```diff
@@ -1,13 +1,76 @@
 package org.oppia.app.parser
 
+import org.oppia.app.customview.interaction.FractionInputInteractionView
 import org.oppia.app.model.Fraction
+import org.oppia.app.topic.FractionParsingErrors
 import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Integer.parseInt
+import java.util.regex.Pattern
 
 /** This class contains method that helps to parse string to fraction. */
 class StringToFractionParser {
   private val wholeNumberOnlyRegex = """^-? ?(\d+)$""".toRegex()
+  private val partialWholeNumberOnlyRegex = """^-? ?$""".toRegex()
   private val fractionOnlyRegex = """^-? ?(\d+) ?/ ?(\d+)$""".toRegex()
+  private val partialFractionOnlyRegex = """^-? ?(\d+) ?/ ?$""".toRegex()
   private val mixedNumberRegex = """^-? ?(\d+) (\d+) ?/ ?(\d+)$""".toRegex()
+  private val partialMixedNumberRegexFirst = """^-? ?(\d+) (\d+) ?$""".toRegex()
+  private val partialMixedNumberRegexSecond = """^-? ?(\d+) (\d+) ?/ ?$""".toRegex()
+  private val invalidCharsRegex = """^[\d\s/-]+$""".toRegex()
+  /**
+   * This method helps to validate the inputText and return [FractionParsingErrors]
+   * This called on text change.
+   * @param inputText is the user input in the [FractionInputInteractionView]
+   * @return enum [FractionParsingErrors]
+   */
+  fun checkForErrors(inputText: String): FractionParsingErrors {
```

Comment:
I'm basing the following on a suggestion I made below regarding error categories, but in general I think we can simplify this parsing. See the following code for a sample (note that you will likely need to change it a bit to make it work--I didn't try to run it. Please use it as a reference for what I think can be a simpler approach):

```kotlin
fun getRealTimeError(text: String): FractionParsingError {
  return when {
    text.matches(invalidCharsRegex) -> FractionParsingError.INVALID_CHARS
    text.matches(invalidFractionRegex) -> FractionParsingError.INVALID_FORMAT
    else -> FractionParsingError.VALID
  }
}

fun getSubmitTimeError(text: String): FractionParsingError {
  // No need to check for real-time errors since the following logically include them.
  val fraction = parseFraction(text)
  return when {
    fraction == null -> FractionParsingError.INVALID_FORMAT
    fraction.denominator == 0 -> FractionParsingError.DIVISION_BY_ZERO
    else -> FractionParsingError.VALID
  }
}

fun getFractionFromString(text: String): Fraction {
  return parseFraction(text) ?: ?: throw IllegalArgumentException("Incorrectly formatted fraction: $text")
}

  private fun parseFraction(text: String): Fraction? {
    // Normalize whitespace to ensure that answer follows a simpler subset of possible patterns.
    val inputText: String = text.normalizeWhitespace()
    return parseMixedNumber(inputText)
      ?: parseFraction(inputText)
      ?: parseWholeNumber(inputText)
  }
```

Three things to note:
1. I renamed the enum to FractionParsingError (enums are usually not plural)
2. I added a check for a bad fraction pattern per Rajat's email (generally meaning we need to check for things like '-' being anywhere but the front, or '//' occurring; we could split this into multiple checks & patterns if that's easier)
3. This leverages existing patterns + a new one for bad fractions and the existing invalid chars pattern you added; the others new ones aren't needed

Reply:
@BenHenning Could you please tell me what invalidFractionRegex here... Iam not able to find a single regex that matches all our requirements.

Reply:
That's partly why I was thinking it might be easier to split it up. I think the examples you mentioned in your later comment could work fine as checks, and no special pattern would be needed:

```kotlin
fun getRealTimeError(text: String): FractionParsingError {
  val normalized = text.normalizeWhitespace()
  return when {
    normalized.matches(invalidCharsRegex) -> FractionParsingError.INVALID_CHARS
    normalized.startsWith("/") -> FractionParsingError.INVALID_FORMAT
    normalized.count { it == '/' } > 1 -> FractionParsingError.INVALID_FORMAT
    normalized.indexOf('-') > 0 -> FractionParsingError.INVALID_FORMAT
    else -> FractionParsingError.VALID // Requires further validation to be sure.
  }
}
```

Reply:
This is not enough there are cases where partial data like  3 3 ,3 3/ would be failing...

Reply:
Can you clarify whether those would fail due to the realtime or submit time checks? Unless I made a mistake, the realtime check code above should return VALID for those two examples, as expected. The submit-time check should fail for both because neither can be successfully parsed into a fraction.

Reply:
@BenHenning Ya it is working perfectly for 3 3 ,3 3/...  there is a small correction in the previous comment its not showing errors in  33 33 33, 3/ 66 99, 55 6 9/ 88 5 and so on...

Reply:
Done.